File: Gemini_Pro/Python/Factory_Method/Round2/cafe.py
from abc import ABC, abstractmethod

class MenuItem(ABC):
    @abstractmethod
    def prepare(self):
        pass

class Beverage(MenuItem):
    pass

class Coffee(Beverage):
    def __init__(self, type):
        self.type = type

    def prepare(self):
        return f"Preparing {self.type} coffee"

class Tea(Beverage):
    def __init__(self, type):
        self.type = type

    def prepare(self):
        return f"Preparing {self.type} tea"

class ColdBeverage(Beverage):
    def __init__(self, type):
        self.type = type

    def prepare(self):
        return f"Preparing {self.type}"

class Pastry(MenuItem):
    pass

class Bread(Pastry):
    def __init__(self, type):
        self.type = type

    def prepare(self):
        return f"Preparing {self.type} bread"

class Cake(Pastry):
    def __init__(self, type):
        self.type = type

    def prepare(self):
        return f"Preparing {self.type} cake"

class Cookie(Pastry):
    def __init__(self, type):
        self.type = type

    def prepare(self):
        return f"Preparing {self.type} cookies"

class Snack(MenuItem):
    pass

class Sandwich(Snack):
    def __init__(self, type):
        self.type = type

    def prepare(self):
        return f"Preparing {self.type} sandwich"

class Pasta(Snack):
    def __init__(self, type):
        self.type = type

    def prepare(self):
        return f"Preparing {self.type} pasta"

class Salad(Snack):
    def __init__(self, type):
        self.type = type

    def prepare(self):
        return f"Preparing {self.type} salad"

class Cafe:
    def create_beverage(self, type):
        if type == "Espresso":
            return Coffee(type)
        elif type == "Cappuccino":
            return Coffee(type)
        elif type == "Latte":
            return Coffee(type)
        elif type == "Mocha":
            return Coffee(type)
        elif type == "Americano":
            return Coffee(type)
        elif type == "Black Tea":
            return Tea(type)
        elif type == "Green Tea":
            return Tea(type)
        elif type == "Oolong Tea":
            return Tea(type)
        elif type == "Lemon Tea":
            return Tea(type)
        elif type == "Fruit Smoothies":
            return ColdBeverage(type)
        elif type == "Milkshakes":
            return ColdBeverage(type)
        elif type == "Fruit Juices":
            return ColdBeverage(type)
        else:
            return None

    def create_pastry(self, type):
        if type == "Croissant":
            return Bread(type)
        elif type == "Bun":
            return Bread(type)
        elif type == "Custard Bread":
            return Bread(type)
        elif type == "Chocolate Cake":
            return Cake(type)
        elif type == "Strawberry Cake":
            return Cake(type)
        elif type == "Coconut Cake":
            return Cake(type)
        elif type == "Chocolate Chip Cookies":
            return Cookie(type)
        elif type == "Oat Cookies":
            return Cookie(type)
        else:
            return None

    def create_snack(self, type):
        if type == "Chicken Sandwich":
            return Sandwich(type)
        elif type == "Tuna Sandwich":
            return Sandwich(type)
        elif type == "Spaghetti":
            return Pasta(type)
        elif type == "Penne":
            return Pasta(type)
        elif type == "Vegetable Salad":
            return Salad(type)
        elif type == "Chicken Salad":
            return Salad(type)
        else:
            return None

File: Gemini_Pro/Java/Factory_Method/Round2/Cafe.java
public class Cafe {

    public interface MenuItem {
        String prepare();
    }

    public static class Coffee implements MenuItem {
        private String type;

        public Coffee(String type) {
            this.type = type;
        }

        @Override
        public String prepare() {
            return "Preparing " + type + " coffee";
        }
    }

    public static class Tea implements MenuItem {
        private String type;

        public Tea(String type) {
            this.type = type;
        }

        @Override
        public String prepare() {
            return "Preparing " + type + " tea";
        }
    }

    public static class ColdBeverage implements MenuItem {
        private String type;

        public ColdBeverage(String type) {
            this.type = type;
        }

        @Override
        public String prepare() {
            return "Preparing " + type;
        }
    }

    public static class Bread implements MenuItem {
        private String type;

        public Bread(String type) {
            this.type = type;
        }

        @Override
        public String prepare() {
            return "Preparing " + type + " bread";
        }
    }

    public static class Cake implements MenuItem {
        private String type;

        public Cake(String type) {
            this.type = type;
        }

        @Override
        public String prepare() {
            return "Preparing " + type + " cake";
        }
    }

    public static class Cookie implements MenuItem {
        private String type;

        public Cookie(String type) {
            this.type = type;
        }

        @Override
        public String prepare() {
            return "Preparing " + type + " cookies";
        }
    }

    public static class Sandwich implements MenuItem {
        private String type;

        public Sandwich(String type) {
            this.type = type;
        }

        @Override
        public String prepare() {
            return "Preparing " + type + " sandwich";
        }
    }

    public static class Pasta implements MenuItem {
        private String type;

        public Pasta(String type) {
            this.type = type;
        }

        @Override
        public String prepare() {
            return "Preparing " + type + " pasta";
        }
    }

    public static class Salad implements MenuItem {
        private String type;

        public Salad(String type) {
            this.type = type;
        }

        @Override
        public String prepare() {
            return "Preparing " + type + " salad";
        }
    }

    public static class CafeFactory {
        public MenuItem createMenuItem(String category, String type) {
            if ("Coffee".equals(category)) {
                switch (type) {
                    case "Espresso":
                    case "Cappuccino":
                    case "Latte":
                    case "Mocha":
                    case "Americano":
                        return new Coffee(type);
                    default:
                        return null; // หรือ throw exception
                }
            } else if ("Tea".equals(category)) {
                switch (type) {
                    case "Black Tea":
                    case "Green Tea":
                    case "Oolong Tea":
                    case "Lemon Tea":
                        return new Tea(type);
                    default:
                        return null; // หรือ throw exception
                }
            } else if ("ColdBeverage".equals(category)) {
                switch (type) {
                    case "Fruit Smoothies":
                    case "Milkshakes":
                    case "Fruit Juices":
                        return new ColdBeverage(type);
                    default:
                        return null; // หรือ throw exception
                }
            } else if ("Bread".equals(category)) {
                switch (type) {
                    case "Croissant":
                    case "Bun":
                    case "Custard Bread":
                        return new Bread(type);
                    default:
                        return null; // หรือ throw exception
                }
            } else if ("Cake".equals(category)) {
                switch (type) {
                    case "Chocolate Cake":
                    case "Strawberry Cake":
                    case "Coconut Cake":
                        return new Cake(type);
                    default:
                        return null; // หรือ throw exception
                }
            } else if ("Cookie".equals(category)) {
                switch (type) {
                    case "Chocolate Chip Cookies":
                    case "Oat Cookies":
                        return new Cookie(type);
                    default:
                        return null; // หรือ throw exception
                }
            } else if ("Sandwich".equals(category)) {
                switch (type) {
                    case "Chicken Sandwich":
                    case "Tuna Sandwich":
                        return new Sandwich(type);
                    default:
                        return null; // หรือ throw exception
                }
            } else if ("Pasta".equals(category)) {
                switch (type) {
                    case "Spaghetti":
                    case "Penne":
                        return new Pasta(type);
                    default:
                        return null; // หรือ throw exception
                }
            } else if ("Salad".equals(category)) {
                switch (type) {
                    case "Vegetable Salad":
                    case "Chicken Salad":
                        return new Salad(type);
                    default:
                        return null; // หรือ throw exception
                }
            } else {
                return null; // หรือ throw exception
            }
        }
    }
}